docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+
+EnTT allows polymorphically accessing components, to allow for deleting, copying, and listing all components an entity has. These features should never be used outside of debugging, since they potentially modify components and calls functions outside of their respective systems.
+
+In short, **components must only be modified by a few select systems**. There's a few other practical reasons why:
+
+* Components may need to be synced with external references: pointers, GPU resourses, or physics engine objects. Some of these require specific threads.
+* Multiple Systems that modify different components can run in parallel without mutexes.
+
+There's a few important operations that need to be addressed however:
+
+* **Deleting entire entities**: Deleting an entity means modifying every component container the entity is associated with. Entities can be marked for deletion with a component, or external queue. 'Deleter' Systems can then delete their respective components near the end of the frame.
+* **Copying entire entities**: It may seem convenient to iterate all of an entity's components and copy them into another entity, but this breaks the rules. Copies of an entity should be created by the same system as the original entity. A dedicated 'mark for copy' and copy systems may be practical depending on the case.
+
+### A solution to all world state?
+
+Entities and Components are powerful, but cannot solve every single storage use case alone. EnTT specifically, only allows one component of a unique type per entity, and only one container of these components (pool) per world (registry). One case may require a variable number of the same component for an entity, or the problem may not be well suited for entities and components at all.
+
+EnTT features customizability and future versions can be more versitile, but either way, ECS is best when combined with other data structures besides just components. For example, queues can be used to pass messages between systems.
+
+## Universe rundown
+
+Universe is designed to support many different kinds of Satellites: Planets, Vehicles, Stars, etc... This can be achieved efficiently through Entities and Components, where Satellites are Entities. When considering keeping groups of Satellites in an N-body simulation, patched conics simulation, or just landed on a planet, it also becomes apparent that some weird scheme is needed to position Satellites.
+
+See [test_application/universes/simple.cpp](https://github.com/TheOpenSpaceProgram/osp-magnum/blob/master/src/test_application/universes/simple.cpp) for how a simple universe can be setup.
+
+### Components
+
+Since Universe uses ECS, Satellites are compositions of Components. Universe Components are simple structs prefixed with **UComp**.
```

Comment:
I still think we should replace UComp with something like Universe::Components::, which can then be "using namespace UComp = Universe::Components;"

Reply:
I tried this before and found it a headache to have many namespaces with the same name as they conflict.

Reply:
That'd be the reason why you guys should stop doing `using namespace Magnum;` at the top of every file, and similar.

You're making a namespace style system by using the names of the classes, instead of using the actual namespace feature of the language.

Reply:
From observation, we are using namespaces to determine which 'module' a type is *part of*, and already using prefixes to determine what a type *is*. "I" for interfaces, "E" for enums, so why not use prefixes for components too?

Reply:
UComp doesn't say what UCompBlah is, in terms of being an enum, pure-abstract-class, or similar.

UCompBlah here could be a primitive type, an enum, a struct, and so on.

Reply:
UComp means it's a component for Universe, which isn't that complicated. Usually it's a struct, but can be any complete and constructible type.